src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -4538,16 +4538,64 @@ void MacroAssembler::load_method_holder(Register holder, Register method) {
   movptr(holder, Address(holder, ConstantPool::pool_holder_offset_in_bytes())); // InstanceKlass*
 }
 
-void MacroAssembler::load_klass(Register dst, Register src, Register tmp) {
+void MacroAssembler::load_klass(Register dst, Register src, Register tmp, bool null_check_src) {
   assert_different_registers(src, tmp);
   assert_different_registers(dst, tmp);
 #ifdef _LP64
-  if (UseCompressedClassPointers) {
-    movl(dst, Address(src, oopDesc::klass_offset_in_bytes()));
-    decode_klass_not_null(dst, tmp);
-  } else
+  assert(UseCompressedClassPointers, "expect compressed class pointers");
+
+  Label slow, done;
+  if (null_check_src) {
+    null_check(src, oopDesc::mark_offset_in_bytes());
+  }
+  movq(tmp, Address(src, oopDesc::mark_offset_in_bytes()));
+  // NOTE: While it would seem nice to use xorb instead (for which we don't have an encoding in our assembler),
+  // the encoding for xorq uses the signed version (0x81/6) of xor, which encodes as compact as xorb would,
+  // and does't make a difference performance-wise.
```

Comment:
I can give you `xorb`, if you want, in a subsequent PR.

Reply:
I actually have it, locally, but as comment says, it doesn't look any better in terms of code size or performance, so why bother?